File: dynamic_programming/solving_questions_with_brainpower.java
class SolutionTopDown
{
    Map<Integer, Long> memo = new HashMap<>();

    public long dp(int i, int[][] questions)
    {
        if (i >= questions.length)
        {
            return 0;
        }

        if (memo.containsKey(i))
        {
            return memo.get(i);
        }

        int next_idx = i + questions[i][1] + 1;
        memo.put(
            i,
            Math.max(questions[i][0] + dp(next_idx, questions),
                     dp(i + 1, questions)));
        return memo.get(i);
    }

    public long mostPoints(int[][] questions)
    {
        return dp(0, questions);
    }
}

public long mostPointsBottomUp(int[][] questions)
{
    int questions_length = questions.length;

    // + 1 to avoid out of bounds
    long[] dp = new long[questions_length + 1];

    for (int i = questions_length - 1; i >= 0; i--)
    {
        int next_idx = i + questions[i][1] + 1;

        // Need to make sure we don't go out of bounds
        dp[i] =
            Math.max(questions[i][0] + dp[Math.min(next_idx, questions_length)],
                     dp[i + 1]);
    }

    return dp[0];
}

File: dynamic_programming/solving_questions_with_brainpower.py
from functools import cache


def mostPointsTopDown(questions: list[list[int]]) -> int:
    @cache
    def dp(i: int) -> int:
        if i >= len(questions):
            return 0

        next_index = i + questions[i][1] + 1
        return max(questions[i][0] + dp(next_index), dp(i + 1))

    return dp(0)


def mostPointsBottomUp(questions: list[list[int]]) -> int:
    questions_len = len(questions)

    # + 1 to avoid out of bounds
    dp = [0] * (questions_len + 1)

    for i in range(questions_len - 1, -1, -1):
        # Need to make sure we don't go out of bounds
        next_index = min(i + questions[i][1] + 1, questions_len)

        dp[i] = max(questions[i][0] + dp[next_index], dp[i + 1])

    return dp[0]
